src/carousel/index.ts

```diff
@@ -0,0 +1,348 @@
+import { tag, utils, Tag } from '@storefront/core';
+
+const DEFAULT_SETTINGS = {
+  speed: 0,
+  slidesToShow: 1,
+  slidesToScroll: 1
+};
+
+@tag('gb-carousel', require('./index.html'), require('./index.css'))
+class Carousel {
+  refs: {
+    carouselwrap: HTMLDivElement,
+    track: HTMLDivElement
+  };
+
+  props: Carousel.Props = <any>{ settings: DEFAULT_SETTINGS };
+  animationEndCallback: NodeJS.Timer | null = null;
+
+  state: Carousel.State = {
+    currentSlide: 0,
+    transitioning: false,
+    touchObject: {
+      startX: null,
+      startY: null,
+      curX: null,
+      curY: null
+    }
+  };
+
+  onMount() {
+    utils.WINDOW().addEventListener('resize', this.updateWindow);
+  }
+
+  onUnMount() {
+    utils.WINDOW().removeEventListener('resize', this.updateWindow);
+  }
+
+  updateWindow = () => {
+    this.update();
+  }
+
+  moveNext = () => {
+    this.slideHandler(
+      this.state.currentSlide +
+      (this.props.settings.slidesToScroll || DEFAULT_SETTINGS.slidesToScroll)
+    );
+  }
+
+  movePrevious = () => {
+    this.slideHandler(
+      this.state.currentSlide -
+      (this.props.settings.slidesToScroll || DEFAULT_SETTINGS.slidesToScroll)
```

Comment:
I wrote this because if client only passes in, say, speed in settings, the whole props.settings will only have speed value but not anything else, so  all three properties in settings need to have individual fallback values. But you are right, I really it is not a good patter. Since I am modifying settings, I move settings under state and initialize their default value from there.